Google.Api.Generator/Utils/ResourcePattern.cs

```diff
@@ -16,107 +16,154 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
-using System.Text;
 
 namespace Google.Api.Generator.Utils
 {
     internal class ResourcePattern
     {
-        public class Segment
+        public abstract class Segment
         {
-            public Segment(string segment)
+            public static Segment Create(string segment) =>
+                segment.Contains('{') ? (Segment)new ResourceIdSegment(segment) : new CollectionIdentifierSegment(segment);
+
+            public abstract IReadOnlyList<string> ParameterNames { get; }
+            public abstract IReadOnlyList<string> ParameterNamesWithSuffix { get; }
+            public abstract IReadOnlyList<char> Separators { get; }
+            public abstract string PathTemplateString { get; }
+            public abstract string Expand(IEnumerable<string> parameters);
+
+            public int ParameterCount => ParameterNames.Count;
+            public bool IsComplex => Separators.Count > 0;
+
+        }
+
+        private class CollectionIdentifierSegment : Segment
+        {
+            public CollectionIdentifierSegment(string segment)
+            {
+                bool valid = segment != "" &&
```

Comment:
It feels like it might be worth extracting this validation, and the one for the separator characters, so you can reuse them. Otherwise if anything changes, it'll be really easy to only update one...

Reply:
None of the validations are quite the same, so I'll leave it as-is for now.
Very happy to extract in a future PR.